codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java

```diff
@@ -681,6 +680,15 @@ private LastHttpContent readTrailingHeaders(ByteBuf buffer) {
             this.trailer = null;
             return trailer;
         }
+        // We have received the empty line which signals the trailer is complete. In this case we need to check if
```

Comment:
This flow is a bit convoluted. How about we remove the surrounding if and make it clear why the condition is there:

```
if (line.length() == 0 && this.trailer == null) {
  return LastHttpContent.EMPTY_LAST_CONTENT; // optimization
}
// everything from the if today
LastHttpContent trailer = this.trailer;
if (trailer == null) {
...
return trailer;
```

Alternatively the if could be changed to `line.length() > 0 || this.trailer != null`, but it's non-obvious why it would be that way. (It's still more obvious than this code, in my mind, though.)